internal/api/v1/application/exec.go

```diff
@@ -0,0 +1,102 @@
+package application
+
+import (
+	"net/http"
+	"net/http/httputil"
+	"time"
+
+	"github.com/epinio/epinio/helpers/kubernetes"
+	"github.com/epinio/epinio/internal/application"
+	"github.com/epinio/epinio/internal/namespaces"
+	apierror "github.com/epinio/epinio/pkg/api/core/v1/errors"
+	"github.com/gin-gonic/gin"
+	v1 "k8s.io/api/core/v1"
+	thekubernetes "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+)
+
+func (hc Controller) Exec(c *gin.Context) apierror.APIErrors {
+	ctx := c.Request.Context()
+	namespace := c.Param("namespace")
+	appName := c.Param("app")
+
+	cluster, err := kubernetes.GetCluster(ctx)
+	if err != nil {
+		return apierror.InternalError(err)
+	}
+
+	exists, err := namespaces.Exists(ctx, cluster, namespace)
+	if err != nil {
+		return apierror.InternalError(err)
+	}
+
+	if !exists {
+		return apierror.NamespaceIsNotKnown(namespace)
+	}
+
+	app, err := application.Lookup(ctx, cluster, namespace, appName)
+	if err != nil {
+		return apierror.InternalError(err)
+	}
+
+	if app == nil {
+		return apierror.AppIsNotKnown(appName)
+	}
+
+	// app exists but has no workload to connect to
+	if app.Workload == nil {
+		return apierror.NewAPIError("Cannot connect to application without workload",
+			"", http.StatusBadRequest)
+	}
+
+	// TODO: Find podName from application and params (e.g. instance number etc).
+	// The application may have more than one pods.
+	podNames, err := application.NewWorkload(cluster, app.Meta).PodNames(ctx)
+	if err != nil {
+		return apierror.InternalError(err)
+	}
+	if len(podNames) < 1 {
+		return apierror.NewAPIError("couldn't find any Pods to connect to",
+			"", http.StatusBadRequest)
+	}
+
+	proxyRequest(c.Writer, c.Request, podNames[0], namespace, appName, cluster.Kubectl)
+
+	return nil
+}
+
+func proxyRequest(rw http.ResponseWriter, req *http.Request, podName, namespace, container string, client thekubernetes.Interface) {
+	// https://github.com/kubernetes/kubectl/blob/2acffc93b61e483bd26020df72b9aef64541bd56/pkg/cmd/exec/exec.go#L352
+	attachURL := client.CoreV1().RESTClient().
+		Post().
+		Namespace(namespace).
+		Resource("pods").
+		Name(podName).
+		SubResource("exec").
+		VersionedParams(&v1.PodExecOptions{
+			Stdin:     true,
+			Stdout:    true,
+			Stderr:    true,
+			TTY:       true,
+			Container: container,
+			// TODO: What if the container doesn't have bash?
```

Comment:
I guess this depends on the builder image and what kind of base image is used for the applications. Maybe we should leave this as it is until this is a problem (e.g. someone wants to use a custom builder image that produces images without bash?)

Reply:
Removing `TODO`. We will fix this if it's ever a problem